client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -83,116 +88,176 @@
+     *                          This is used if the controller endpoints are static and can be directly accessed.
+     *                      2. pravega://ip1:port1,ip2:port2,...
+     *                          This is used to autodiscovery the controller endpoints from an initial controller list.
+     * @param config        The configuration for this client implementation.
+     */
+    public ControllerImpl(final URI controllerURI, final ControllerImplConfig config) {
+        this(controllerURI, config, Executors.newSingleThreadScheduledExecutor());
+        this.internalExecutor.set(this.executor);
+    }
+
+    /**
+     * Creates a new instance of the Controller client class.
+     *
+     * @param controllerURI The controller rpc URI. This can be of 2 types
+     *                      1. tcp://ip1:port1,ip2:port2,...
+     *                          This is used if the controller endpoints are static and can be directly accessed.
+     *                      2. pravega://ip1:port1,ip2:port2,...
+     *                          This is used to autodiscovery the controller endpoints from an initial controller list.
+     * @param config        The configuration for this client implementation.
+     * @param executor      The executor service to be used for handling retries.
+     */
+    public ControllerImpl(final URI controllerURI, final ControllerImplConfig config,
+                          final ScheduledExecutorService executor) {
         this(NettyChannelBuilder.forTarget(controllerURI.toString())
                 .nameResolverFactory(new ControllerResolverFactory())
                 .loadBalancerFactory(RoundRobinLoadBalancerFactory.getInstance())
                 .keepAliveTime(DEFAULT_KEEPALIVE_TIME_MINUTES, TimeUnit.MINUTES)
-                .usePlaintext(true));
+                .usePlaintext(true), config, executor);
         log.info("Controller client connecting to server at {}", controllerURI.getAuthority());
     }
 
     /**
      * Creates a new instance of the Controller client class.
      *
      * @param channelBuilder The channel builder to connect to the service instance.
+     * @param config         The configuration for this client implementation.
+     * @param executor       The executor service to be used internally.
      */
     @VisibleForTesting
-    public ControllerImpl(ManagedChannelBuilder<?> channelBuilder) {
+    public ControllerImpl(ManagedChannelBuilder<?> channelBuilder, final ControllerImplConfig config,
+                          final ScheduledExecutorService executor) {
         Preconditions.checkNotNull(channelBuilder, "channelBuilder");
 
+        this.executor = executor;
+        this.retryConfig = Retry.withExpBackoff(config.getInitialBackoffMillis(), config.getBackoffMultiple(),
+                config.getRetryAttempts(), config.getMaxBackoffMillis())
+                .retryingOn(StatusRuntimeException.class)
+                .throwingOn(Exception.class);
```

Comment:
No, the APIs don't internally throw RetriesExhaustedException. RetriesExhaustedException is thrown by Retry class on these configured conditions. So the current configuration means - retry only on gRPC exceptions and terminate for all other exceptions.